src/cc/perf_reader.c

```diff
@@ -70,7 +73,12 @@ struct perf_reader * perf_reader_new(perf_reader_cb cb,
 void perf_reader_free(void *ptr) {
   if (ptr) {
     struct perf_reader *reader = ptr;
-    while (!__sync_bool_compare_and_swap(&reader->rb_use_state, RB_NOT_USED, RB_USED_IN_MUNMAP));
+    pid_t tid = syscall(__NR_gettid);
+    while (!__sync_bool_compare_and_swap(&reader->rb_use_state, RB_NOT_USED, RB_USED_IN_MUNMAP)) {
+      // If the same thread, it is called from call back handler, no locking needed
+      if (tid == reader->rb_read_tid)
```

Comment:
We'll read a read barrier here, right?

Reply:
__sync_bool_compare_and_swap will act as a memory barrier.
https://gcc.gnu.org/onlinedocs/gcc/_005f_005fsync-Builtins.html
Quote
```
In most cases, these built-in functions are considered a full barrier. That is, no memory operand is moved across the operation, either forward or backward. Further, instructions are issued as necessary to prevent the processor from speculating loads across the operation and from queuing stores after the operation. 
```
`__sync_bool_compare_and_swap` is a full barrier.